File: bindings/swift/Sources/RapidlyEngine/RapidlyEngine.swift
//
//  RapidlyEngine.swift
//  RapidlyEngine
//
//  Copyright (c) 2026 Rapidly Labs AS.
//
//  You are not allowed to use, distribute or modify this code without
//  a written permission from Rapidly Labs AS.
//
//  Swift-native API for the Rapidly SDK. Ships as a SwiftPM source
//  target on top of the binary RapidlyEngineC xcframework. Customers
//  who add the Rapidly SDK via SwiftPM get this Swift class plus the
//  re-exported C API in a single `import RapidlyEngine`.
//

import Foundation
@_exported import RapidlyEngineC
import RapidlyEngineC.Private  // internal Obj-C++ adapter; not re-exported to consumers

/// Swift-native audio processor wrapping the Rapidly C engine.
///
/// ## Threading
///
/// A `RapidlyEngine` instance is **not safe to call concurrently from
/// multiple threads** at any moment. The class follows the same
/// concurrency model as Apple's `AVAudioEngine` and similar real-time
/// audio types: the customer is responsible for serialising access.
///
/// Apps that route audio through a render callback on a high-priority
/// audio thread should follow this exact stop sequence:
///
/// 1. Stop your audio session / unit / render callback first. After
///    this returns, the audio thread is guaranteed not to be inside
///    `addAudio*`, `getAudio*`, `getProcessorInfo()`, etc.
/// 2. Call `close()` on this engine.
/// 3. Release the engine reference.
///
/// **Calling `close()` while another thread is mid-call into the same
/// engine is undefined behaviour** — it can crash, corrupt memory, or
/// behave unpredictably. This is consistent with how every real-time
/// audio SDK on Apple platforms behaves; coordinating thread lifecycle
/// is a customer responsibility, not the SDK's.
///
/// Under Swift 6 strict concurrency, `RapidlyEngine` is intentionally
/// NOT declared `Sendable` for the same reason — sending it across
/// actor boundaries without serialising access is unsafe. If you must
/// share the engine across actors, wrap it in your own
/// `@unchecked Sendable` adapter and serialise access in that wrapper.
///
/// ## Lifetime
///
/// Construction loads the model file and (per the global licence pool
/// at that moment) bakes in the watermarked-or-not state for the lifetime
/// of the instance. Call `RapidlyEngine.addLicense(_:)` BEFORE
/// construction to avoid watermarking.
///
/// For real-time audio (the dominant use case), explicit `close()` is
/// effectively required. ARC-only release is safe only in contexts
/// where no audio thread can possibly still be in flight (e.g.,
/// short-lived offline processing). The `deinit` calls `close()`
/// defensively, but its timing relative to your audio session is not
/// deterministic. Follow the stop sequence above for safe release from
/// real-time audio contexts.
public class RapidlyEngine {
    
    public struct ProcessorInfo {
        public var sampleRate: Double
        public var numOfModelChannels: Int
        public var latencyInSamples: Int
    }

    public enum Parameters: UInt32 {
        /** The maximum attenuation of the unwanted signal in dB. The
             valid value range is <-inf, 0] */
        case maximumAttenuation = 0x0001
        /** The sensitivity of the processing in percent ranging from
             -100% to 100%. 0% is neutral and positive values will increase
             the amount of reduction. */
        case sensitivity = 0x0002
        /** Enables or disables mask frequency extrapolation. When using
             models that aren't trained on the full frequency band, this setting
             determines if the output audio should be band-limited to the
             frequency range of the model or if the high-frequency masks should
             be extrapolated. Settings below 0.5 disables the extrapolaion
             while settings equal or above 0.5 enables it.  */
        case maskExtrapolation = 0x0003
        /** The stem dependent sensitivities of the processing in
             percent ranging from -100% to 100%. 0% is neutral and positive
             values will increase the amount of reduction. You can set
             individual stems by adding the model index to the define, so
             firstModelSensitivity will be the first model,
             firstModelSensitivity + 1 the second, and so forth. */
        case firstModelSensitivity = 0x1000
    }
   
    private var engineAdapter: EngineAdapterObjC

    /**
     * Creates an audio processor, loads a model file and returns a processor instance if successful.
     * - Parameters:
     *     - modelFilepath: Pointer to a string containing the file path of the model file to load.
     *     - numOfChannels: The number of channels in the audio to process.
     *     - sampleRate: The sample rate of the audio to process.
     * - Returns: A valid object on success. Returns `nil` only if the model file is missing or
     *   unreadable. Licence coverage never causes `nil`: a model your licence does not cover (or
     *   any model loaded with no licence) still constructs successfully, but its output is watermarked.
     */
    public init?(modelFilepath: String, numOfChannels: UInt32, sampleRate: Double) {
        guard let adapter = EngineAdapterObjC(modelFilepaths: [modelFilepath], numOfChannels: numOfChannels, sampleRate: sampleRate) else {
            return nil
        }
        self.engineAdapter = adapter
    }
    
    /**
     * Creates a processor that performs stem separation, loads a set of model files and returns a processor instance if successful.
     * The number of stems is derived from modelFilepaths.count -- the array already carries its length.
     * - Parameters:
     *     - modelFilepaths: An array of strings containing the file paths of the model files to load.
     *     - numOfChannels: The number of channels in the audio to process.
     *     - sampleRate: The sample rate of the audio to process.
     * - Returns: A valid object on success. Returns `nil` only if a model file is missing or
     *   unreadable. Licence coverage never causes `nil`: a model your licence does not cover (or
     *   any model loaded with no licence) still constructs successfully, but its output is watermarked.
     */
    public init?(modelFilepaths: [String], numOfChannels: UInt32, sampleRate: Double) {
        guard let adapter = EngineAdapterObjC(modelFilepaths: modelFilepaths, numOfChannels: numOfChannels, sampleRate: sampleRate) else {
            return nil
        }
        self.engineAdapter = adapter
    }
    
    /**
     Adds a license key to the Rapidly SDK. Licence state is global, so this must be called
     BEFORE constructing any `RapidlyEngine` instance. Without a covering licence, the SDK
     operates in a watermarked demo mode that audibly degrades its output, and that state is
     baked into each processor at creation time -- a later licence call does not retroactively
     un-watermark already-created processors.
     - Parameters:
        - licenseString: A string containing a license received from Rapidly.
     - Returns: `true` if the license check succeeded, otherwise `false`.
    */
    public static func addLicense(_ licenseString: String) -> Bool {
        return EngineAdapterObjC.addLicense(licenseString)
    }

    /**
     * Adds floating point PCM encoded audio from separate channels to the Rapidly audio processor.
     * - Parameters:
     *     - pcmChannels: Pointer to an array of channel data pointers, each pointing to sampled PCM values as 32-bit floating point.
     *     - numOfSamples: The number of samples to add
     */
    public func addAudio(pcmChannels: [UnsafeMutableRawPointer?], numOfSamples: UInt32) {
        engineAdapter.addAudio(pcmChannels, numOfSamples:numOfSamples)
    }

    /**
     * Adds floating point PCM encoded audio from a single channel-interleaved buffer to the Rapidly audio processor.
     * - Parameters:
     *     - pcmChannels: Pointer to a 32-bit floating point buffer containing channel-interleaved PCM audio (stereo audio will be in the form "Left Sample 1", "Right Sample 1", "Left Sample 2"...).
     *     - numOfSamples: The number of samples to add
     */
    public func addAudioInterleaved(pcmChannels: UnsafeMutableRawPointer?, numOfSamples: UInt32) {
        engineAdapter.addAudioInterleaved(pcmChannels, numOfSamples:numOfSamples)
    }

    /**
     * Returns the number of samples that are ready after model inference. If the end of the stream has been reached, endOfStream can
     * can be set to true to retrieve the number of remaining samples in the processing queue.
     * - Returns: Number of completed samples.
     */
    public func getNumOfPendingSamples() -> UInt32 {
        return engineAdapter.getNumOfPendingSamples()
    }

    /**
     * Gets floating point PCM encoded audio in separate channels after processing. The number of requested samples must be less or equal to the number of available samples as returned by ``getNumOfPendingSamples``.
     * - Parameters:
     *     - pcmChannels: Pointer to an array of channel data pointers, each receiving sampled PCM values as 32 bit floating point.
     *     - numOfSamples: The number of samples to retrieve
     */
    @discardableResult
    public func getAudio(pcmChannels: [UnsafeMutableRawPointer?], numOfSamples: UInt32) -> Bool {
        return engineAdapter.getAudio(pcmChannels, numOfSamples: numOfSamples)
    }

    /**
     * Gets floating point PCM encoded audio in a single channel-interleaved buffer after processing. The number of requested samples must be less or equal to the number of available samples as returned by ``getNumOfPendingSamples``.
     * - Parameters:
     *     - pcmChannels: Pointer to an array of channel data pointers, each receiving sampled PCM values as 32 bit floating point.
     *     - numOfSamples: The number of samples to retrieve
     */
    @discardableResult
    public func getAudioInterleaved(pcmChannels: UnsafeMutableRawPointer?, numOfSamples: UInt32) -> Bool {
        return engineAdapter.getAudioInterleaved(pcmChannels, numOfSamples:numOfSamples)
    }

    /**
     * Resets the processor state and clears all delay lines.
     */
    public func resetProcessorState() {
        engineAdapter.resetProcessorState()
    }

    /**
     * Get information about the Rapidly processor.
     - Returns: A RapidlyEngine.ProcessorInfo structure
     */
    public func getProcessorInfo() -> ProcessorInfo {
        let info = engineAdapter.getProcessorInfo()
        return ProcessorInfo(sampleRate: info.sampleRate, numOfModelChannels: Int(info.numOfModelChannels), latencyInSamples: Int(info.latencyInSamples))
    }
    
    /**
     * Returns a parameter value
     * - Parameters:
     *     - for: One of the RapidlyEngine.Parameters.
     * - Returns: The current parameter value.
     */
    public func getParameterValue(for param: Parameters) -> Float {
        return engineAdapter.getParameterValue(for: param.rawValue)
    }

    /**
     * Sets a parameter value
     * - Parameters:
     *     - for: One of the RapidlyEngine.Parameters.
     *     - to: The new parameter value.
     */
    public func setParameter(for param: Parameters, to value: Float) {
        engineAdapter.setParameterFor(param.rawValue, to: value)
    }

    /**
     Releases the underlying processor, freeing its resources. Safe to call
     multiple times. After `close`, all audio methods become no-ops and read
     methods return false / zero. For real-time audio, calling this from your
     audio-stop callback BEFORE releasing references to the engine is
     effectively required: it avoids a race where the audio thread is still
     inside `addAudio` / `getAudio` when the processor is freed. The engine's
     `deinit` also calls `close` defensively, but its timing relative to your
     audio session is not deterministic, so do not rely on it for real-time
     contexts.
    */
    public func close() {
        engineAdapter.close()
    }

}

File: bindings/kotlin/src/main/kotlin/io/rapidly/engine/RapidlyEngine.kt
/*

This file is part of the Rapidly SDK for cross-platform model inference.
Copyright (c) 2026 Rapidly Labs AS.

You are not allowed to use, distribute or modify this code without
a written permission from Rapidly Labs AS.

*/

//
//  RapidlyEngine.kt
//  RapidlyEngine -- Kotlin / Android binding
//
//  Kotlin-native API for the Rapidly SDK. Ships as the rapidly-engine
//  Android library (.aar) with Maven coordinates io.rapidly:engine.
//
//  This is the idiomatic class layer of the binding. It holds no logic of
//  its own: every method is a one-line delegation to the JNI shim
//  (libRapidlyEngineJni.so), which marshals to the stable public C API,
//  which runs the C++ engine. It mirrors the Swift binding's RapidlyEngine
//  surface and semantics one-to-one, expressed in Kotlin idiom (see the
//  SDK language-bindings architecture).
//

package io.rapidly.engine

import java.nio.FloatBuffer

/**
 * Kotlin-native audio processor wrapping the Rapidly C engine.
 *
 * ## Threading (read this before shipping)
 *
 * A [RapidlyEngine] instance is **not safe to call concurrently from
 * multiple threads** at any moment. It follows the same concurrency model
 * as Android's real-time audio types (`AudioTrack`, `AudioRecord`, an Oboe
 * stream callback): the caller is responsible for serialising access.
 *
 * Apps that feed audio from a real-time callback thread should follow this
 * exact stop sequence:
 *
 * 1. Stop your audio stream / callback first. After this returns, the audio
 *    thread is guaranteed not to be inside `addAudio*`, `getAudio*`,
 *    [getProcessorInfo], etc.
 * 2. Call [close] on this engine.
 * 3. Release the engine reference.
 *
 * **Calling [close] while another thread is mid-call into the same engine
 * is undefined behaviour.** It can crash, corrupt memory, or behave
 * unpredictably. This is consistent with how every real-time audio API on
 * Android behaves; coordinating thread lifecycle is a caller responsibility,
 * not the SDK's.
 *
 * ## Lifetime
 *
 * Construction loads the model file and (per the global licence pool at
 * that moment) bakes in the watermarked-or-not state for the lifetime of
 * the instance. Call [addLicense] BEFORE construction to avoid watermarking.
 *
 * [RapidlyEngine] implements [AutoCloseable]. For real-time audio the
 * dominant use case explicit [close] is effectively required: it frees the
 * native processor and the model memory it holds. There is no deterministic
 * finalizer (unlike the Swift binding's ARC `deinit`, the JVM offers no
 * equivalent), so a forgotten [close] leaks the native processor until
 * process exit. For short-lived offline processing prefer Kotlin's
 * [use][kotlin.io.use]:
 *
 * ```
 * RapidlyEngine.create(modelPath, numOfChannels = 1, sampleRate = 48000.0)
 *     ?.use { engine -> /* ... */ }
 * ```
 *
 * ## Construction failure
 *
 * The [create] factories return `null` only when a model file is missing
 * or unreadable. Licence coverage never causes a `null` result: a model
 * your licence does not cover (or any model loaded with no licence) still
 * constructs successfully, but its output is watermarked. This mirrors the
 * Swift binding's failable initialiser, where a missing model yields `nil`.
 */
public class RapidlyEngine private constructor(private var handle: Long) : AutoCloseable {

    /** Information about a neural-network audio processor, from [getProcessorInfo]. */
    public data class ProcessorInfo(
        /**
         * Sample rate the model was trained at. The processor automatically
         * converts sample rates to match the model.
         */
        val sampleRate: Double,
        /**
         * The true number of channels used in the processing. The processor
         * automatically converts the channel format to match the model.
         */
        val numOfModelChannels: Int,
        /** The maximum latency of the model, in samples. */
        val latencyInSamples: Int,
    )

    /** Processor parameters readable with [getParameterValue] and writable with [setParameter]. */
    public enum class Parameters(public val value: Int) {
        /** The maximum attenuation of the unwanted signal in dB. The valid value range is <-inf, 0]. */
        MAXIMUM_ATTENUATION(0x0001),

        /**
         * The sensitivity of the processing in percent ranging from -100% to
         * 100%. 0% is neutral and positive values increase the amount of reduction.
         */
        SENSITIVITY(0x0002),

        /**
         * Enables or disables mask frequency extrapolation. When using models
         * that are not trained on the full frequency band, this setting
         * determines if the output audio should be band-limited to the
         * frequency range of the model or if the high-frequency masks should
         * be extrapolated. Settings below 0.5 disable extrapolation, settings
         * equal to or above 0.5 enable it.
         */
        MASK_EXTRAPOLATION(0x0003),

        /**
         * The stem-dependent sensitivities of the processing in percent
         * ranging from -100% to 100%. 0% is neutral and positive values
         * increase the amount of reduction. Address an individual stem by
         * adding its model index to this value, so [FIRST_MODEL_SENSITIVITY]
         * is the first model, `FIRST_MODEL_SENSITIVITY.value + 1` the second,
         * and so forth.
         */
        FIRST_MODEL_SENSITIVITY(0x1000),
    }

    /**
     * Adds floating-point PCM encoded audio from separate channels to the processor.
     *
     * @param pcmChannels One direct [FloatBuffer] per channel, each holding
     *   sampled PCM values. Buffers must be direct (`ByteBuffer.allocateDirect`
     *   based) so the audio crosses to native code without a copy; audio is
     *   read starting at element 0.
     * @param numOfSamples The number of samples to add, per channel.
     */
    public fun addAudio(pcmChannels: Array<FloatBuffer>, numOfSamples: Int) {
        nativeAddAudio(handle, pcmChannels, numOfSamples)
    }

    /**
     * Adds floating-point PCM encoded audio from a single channel-interleaved buffer.
     *
     * @param pcmChannels A direct [FloatBuffer] of channel-interleaved PCM
     *   audio (stereo audio is "Left Sample 1", "Right Sample 1",
     *   "Left Sample 2", ...). Must be direct so the audio crosses to native
     *   code without a copy; audio is read starting at element 0.
     * @param numOfSamples The number of samples to add, per channel.
     */
    public fun addAudioInterleaved(pcmChannels: FloatBuffer, numOfSamples: Int) {
        nativeAddAudioInterleaved(handle, pcmChannels, numOfSamples)
    }

    /**
     * Returns the number of samples ready after model inference.
     *
     * Processing introduces latency, so query this before retrieving audio
     * with [getAudio] / [getAudioInterleaved].
     *
     * @return The number of completed samples.
     */
    public fun getNumOfPendingSamples(): Int = nativeGetNumOfPendingSamples(handle)

    /**
     * Gets floating-point PCM encoded audio in separate channels after processing.
     *
     * [numOfSamples] must be less than or equal to the number of available
     * samples returned by [getNumOfPendingSamples].
     *
     * @param pcmChannels One direct [FloatBuffer] per channel, each receiving
     *   sampled PCM values. Buffers must be direct; audio is written starting
     *   at element 0.
     * @param numOfSamples The number of samples to retrieve, per channel.
     * @return `true` on success, otherwise `false`.
     */
    public fun getAudio(pcmChannels: Array<FloatBuffer>, numOfSamples: Int): Boolean =
        nativeGetAudio(handle, pcmChannels, numOfSamples)

    /**
     * Gets floating-point PCM encoded audio in a single channel-interleaved
     * buffer after processing.
     *
     * [numOfSamples] must be less than or equal to the number of available
     * samples returned by [getNumOfPendingSamples].
     *
     * @param pcmChannels A direct [FloatBuffer] receiving channel-interleaved
     *   PCM audio. Must be direct; audio is written starting at element 0.
     * @param numOfSamples The number of samples to retrieve, per channel.
     * @return `true` on success, otherwise `false`.
     */
    public fun getAudioInterleaved(pcmChannels: FloatBuffer, numOfSamples: Int): Boolean =
        nativeGetAudioInterleaved(handle, pcmChannels, numOfSamples)

    /** Resets the processor state and clears all delay lines. */
    public fun resetProcessorState() {
        nativeResetProcessorState(handle)
    }

    /**
     * Returns information about the Rapidly processor.
     *
     * @return A [ProcessorInfo] describing the processor.
     */
    public fun getProcessorInfo(): ProcessorInfo {
        val info = nativeGetProcessorInfo(handle)
        return ProcessorInfo(
            sampleRate = info[0],
            numOfModelChannels = info[1].toInt(),
            latencyInSamples = info[2].toInt(),
        )
    }

    /**
     * Returns a parameter value.
     *
     * @param parameter One of the [Parameters].
     * @return The current parameter value.
     */
    public fun getParameterValue(parameter: Parameters): Float =
        nativeGetParameterValue(handle, parameter.value)

    /**
     * Sets a parameter value.
     *
     * @param parameter One of the [Parameters].
     * @param value The new parameter value.
     */
    public fun setParameter(parameter: Parameters, value: Float) {
        nativeSetParameterValue(handle, parameter.value, value)
    }

    /**
     * Releases the underlying processor, freeing its resources. Safe to call
     * multiple times. After [close], all audio methods become no-ops and read
     * methods return `false` / zero.
     *
     * For real-time audio, calling this from your audio-stop path BEFORE
     * releasing references to the engine is effectively required: it avoids a
     * race where the audio thread is still inside `addAudio` / `getAudio` when
     * the processor is freed. See the threading notes on [RapidlyEngine].
     */
    public override fun close() {
        if (handle != 0L) {
            nativeDeleteProcessor(handle)
            handle = 0L
        }
    }

    public companion object {

        init {
            // Load the engine library directly FIRST: loading through the JVM
            // (rather than leaving it to the dynamic linker as a NEEDED
            // dependency of the JNI shim) makes the JVM invoke the engine's
            // JNI_OnLoad, which the engine needs to reach the platform
            // networking stack for online licence keys. Offline licence keys
            // work either way. Then load the JNI shim; its engine dependency
            // is already satisfied. The .aar packs both .so side by side.
            System.loadLibrary("RapidlyEngine")
            System.loadLibrary("RapidlyEngineJni")
        }

        /**
         * Adds a licence key to the Rapidly SDK.
         *
         * Licence state is global, so this must be called BEFORE constructing
         * any [RapidlyEngine] instance. Without a covering licence the SDK
         * operates in a watermarked demo mode that audibly degrades its
         * output, and that state is baked into each processor at creation
         * time: a later licence call does not retroactively un-watermark
         * already-created processors.
         *
         * @param licenseString A licence string received from Rapidly.
         * @return `true` if the licence check succeeded, otherwise `false`.
         */
        @JvmStatic
        public fun addLicense(licenseString: String): Boolean = nativeAddLicense(licenseString)

        /**
         * This call is ignored because the SDK only ever talks to Rapidly's
         * production API. It does so only when a subscription key or
         * activation key is in use, for licensing and billing. Licence keys
         * are fully offline. Audio never leaves the device.
         */
        @JvmStatic
        public fun setApiBaseUrl(baseUrl: String): Unit = nativeSetApiBaseUrl(baseUrl)

        /**
         * Starts a usage session for one end user. Subscription deployments
         * only; silently ignored otherwise. Use an opaque identifier (a UUID
         * or hash), never an email or name. Non-blocking: audio is never
         * interrupted while the session is established.
         *
         * @return `true` if the SDK is currently authorized for processing.
         */
        @JvmStatic
        public fun startSession(externalUserId: String): Boolean = nativeStartSession(externalUserId)

        /**
         * Ends the current usage session and reports its duration.
         * Subscription deployments only; silently ignored otherwise.
         */
        @JvmStatic
        public fun stopSession(): Unit = nativeStopSession()

        /**
         * Creates an audio processor and loads a single model file.
         *
         * @param modelFilepath File path of the model file to load.
         * @param numOfChannels The number of channels in the audio to process.
         * @param sampleRate The sample rate of the audio to process.
         * @return A [RapidlyEngine] on success. Returns `null` only if the
         *   model file is missing or unreadable. Licence coverage never causes
         *   `null`: a model your licence does not cover (or any model loaded
         *   with no licence) still constructs successfully, but its output is
         *   watermarked.
         */
        @JvmStatic
        public fun create(
            modelFilepath: String,
            numOfChannels: Int,
            sampleRate: Double,
        ): RapidlyEngine? {
            val handle = nativeCreateProcessor(modelFilepath, numOfChannels, sampleRate)
            return if (handle != 0L) RapidlyEngine(handle) else null
        }

        /**
         * Creates a processor that performs stem separation, loading a set of
         * model files. The number of stems is derived from [modelFilepaths].
         *
         * @param modelFilepaths File paths of the model files to load.
         * @param numOfChannels The number of channels in the audio to process.
         * @param sampleRate The sample rate of the audio to process.
         * @return A [RapidlyEngine] on success. Returns `null` only if a model
         *   file is missing or unreadable. Licence coverage never causes
         *   `null`: a model your licence does not cover (or any model loaded
         *   with no licence) still constructs successfully, but its output is
         *   watermarked.
         */
        @JvmStatic
        public fun create(
            modelFilepaths: List<String>,
            numOfChannels: Int,
            sampleRate: Double,
        ): RapidlyEngine? {
            val handle = nativeCreateStemSeparator(
                modelFilepaths.toTypedArray(), numOfChannels, sampleRate,
            )
            return if (handle != 0L) RapidlyEngine(handle) else null
        }

        // --- JNI shim entry points (libRapidlyEngineJni.so) ---------------
        // @JvmStatic hoists each native method onto RapidlyEngine itself, so
        // the JNI symbol is Java_io_rapidly_engine_RapidlyEngine_native*,
        // matching rapidly_jni.cpp. The processor handle is passed explicitly
        // as a Long, so the native side never holds a back-reference to this
        // object.

        @JvmStatic
        private external fun nativeAddLicense(licenseString: String): Boolean

        @JvmStatic
        private external fun nativeSetApiBaseUrl(baseUrl: String)

        @JvmStatic
        private external fun nativeStartSession(externalUserId: String): Boolean

        @JvmStatic
        private external fun nativeStopSession()

        @JvmStatic
        private external fun nativeCreateProcessor(
            modelFilepath: String, numOfChannels: Int, sampleRate: Double,
        ): Long

        @JvmStatic
        private external fun nativeCreateStemSeparator(
            modelFilepaths: Array<String>, numOfChannels: Int, sampleRate: Double,
        ): Long

        @JvmStatic
        private external fun nativeDeleteProcessor(handle: Long)

        @JvmStatic
        private external fun nativeAddAudio(
            handle: Long, pcmChannels: Array<FloatBuffer>, numOfSamples: Int,
        )

        @JvmStatic
        private external fun nativeAddAudioInterleaved(
            handle: Long, interleavedPCM: FloatBuffer, numOfSamples: Int,
        )

        @JvmStatic
        private external fun nativeGetNumOfPendingSamples(handle: Long): Int

        @JvmStatic
        private external fun nativeGetAudio(
            handle: Long, pcmChannels: Array<FloatBuffer>, numOfSamples: Int,
        ): Boolean

        @JvmStatic
        private external fun nativeGetAudioInterleaved(
            handle: Long, interleavedPCM: FloatBuffer, numOfSamples: Int,
        ): Boolean

        @JvmStatic
        private external fun nativeResetProcessorState(handle: Long)

        @JvmStatic
        private external fun nativeGetProcessorInfo(handle: Long): DoubleArray

        @JvmStatic
        private external fun nativeGetParameterValue(handle: Long, parameterIndex: Int): Float

        @JvmStatic
        private external fun nativeSetParameterValue(
            handle: Long, parameterIndex: Int, parameterValue: Float,
        )
    }
}
